spec/index.bs

```diff
@@ -1191,131 +1212,189 @@ access to it before garbage collection by calling {{GPUBuffer/destroy()}}.
-    const GPUBufferUsageFlags UNIFORM       = 0x0040;
-    const GPUBufferUsageFlags STORAGE       = 0x0080;
-    const GPUBufferUsageFlags INDIRECT      = 0x0100;
-    const GPUBufferUsageFlags QUERY_RESOLVE = 0x0200;
-};
-</script>
+  1. Set |this|.{{[[state]]}} to [=buffer state/destroyed=]
+
+  Issue: Handle error buffers once we have a description of the error monad.
+</div>
 
 ## Buffer Mapping ## {#buffer-mapping}
 
-An application can request to map a {{GPUBuffer}} to get its mapping which is
-an {{ArrayBuffer}} representing the {{GPUBuffer}}'s allocation. Mappings are
-requested asynchronously so that the user agent can ensure the GPU finished
-using the {{GPUBuffer}} before the application gets its mapping. Mappings can
-be requested for reading with {{GPUBuffer/mapReadAsync}} or writing with
-{{GPUBuffer/mapWriteAsync}}. A mapped {{GPUBuffer}} cannot be used by the GPU
-and must be unmapped using {{GPUBuffer/unmap}} before it can be used on the
-[=Queue timeline=].
+An application can request to map a {{GPUBuffer}} so that they can access its
+content via {{ArrayBuffer}}s that represent part of the {{GPUBuffer}}'s
+allocations. Mapping a {{GPUBuffer}} is requested asynchronously with
+{{GPUBuffer/mapAsync}} so that the user agent can ensure the GPU
+finished using the {{GPUBuffer}} before the application can access its content.
+Once the {{GPUBuffer}} is mapped the application can synchronously ask for access
+to ranges of its content with {{GPUBuffer/getMappedRange}}. A mapped {{GPUBuffer}}
+cannot be used by the GPU and must be unmapped using {{GPUBuffer/unmap}} before
+it can be used on the [=Queue timeline=].
+
+Issue(gpuweb/gpuweb#605): Add client-side validation that a mapped buffer can
+  only be unmapped and destroyed on the worker on which it was mapped. Likewise
+  {{GPUBuffer/getMappedRange}} can only be called on that worker.
+
+### <dfn method for=GPUBuffer>mapAsync(start, offset)</dfn> ### {#GPUBuffer-mapAsync}
+
+<div algorithm="GPUBuffer.mapAsync">
+
+  **Arguments:**
+    - {{GPUSize64}} |offset|
+    - {{GPUSize64}} |size|
+
+  **Returns:** {{Promise}}
+
+  Issue(gpuweb/gpuweb#605): Handle error buffers once we have a description of the error monad.
+
+  1. If |size| is 0 and |offset| is less than |this|.{{[[size]]}}:
+
+    1. Set |size| to |this|.{{[[size]]}} - |offset|
 
-Issue: Add client-side validation that a mapped buffer can only be unmapped and destroyed on the worker on which it was mapped.
+  1. If this call doesn't follow [$mapAsync Valid Usage$]:
 
-### {{GPUBuffer/mapReadAsync|GPUDevice.mapReadAsync}} ### {#GPUBuffer-mapReadAsync}
+    1. Record a validation error on the current scope.
+    1. Return [=a promise rejected with=] an {{AbortError}} on the [=Device timeline=].
 
-<div algorithm="GPUBuffer.mapReadAsync">
-    Issue: Handle error buffers once we have a description of the error monad.
+  1. Let |p| be a new {{Promise}}.
+  1. Set |this|.{{[[mapping]]}} to |p|.
```

Comment:
[[mapping]] doesn't say it can be a Promise. Maybe this line is not needed though?

Reply:
It does say that it can be a promise. This line is needed to be able to reject a promise when calling unmap early.